### DISCUSSION TITLE
ESLint

### DISCUSSION THREAD

**Discussion by dbjoEq** [equinor/mad#142] [Ideas]
**The time is now.**

As our software team continues to develop innovative mobile applications, it becomes essential to maintain code quality and consistency throughout our projects. One effective tool that can significantly enhance our development process is ESLint. In this discussion, we will explore why implementing ESLint in our tech stack is a crucial step towards achieving cleaner, more reliable, and maintainable code. Here are som **key benefits** to ESLint:

- Code consistency
- Bug prevention
- TypeScript Support
- Enhanced code reviews

The proposed way to add ESLint to projects is to keep the basic rules in [eslint-config-mad-basic](https://github.com/equinor/mad/blob/main/packages/eslint-config-mad-custom/index.js) and then use that to extend our projects. 

Currently mad-basic is very simple, so the question is rules do we want??

**Comment by dbjoEq** [comment]
**Recommended extensions from chatGPT. 🤖** 

eslint-plugin-react: Provides rules specific to React and JSX.

eslint-plugin-react-native: Contains rules for React Native-specific code.

eslint-plugin-import: Helps with import/export statements and modules.

@typescript-eslint/eslint-plugin: Contains TypeScript-specific ESLint rules.

eslint-plugin-prettier: Allows ESLint to show formatting errors using Prettier.

eslint-config-prettier: Disables ESLint rules that conflict with Prettier.

eslint-plugin-redux-saga: Includes rules for working with Redux Saga.

eslint-plugin-redux-hooks: Contains rules for using React-Redux hooks.

**Comment by dbjoEq** [comment]
**Recommended rules from chatGPT. 🤖** 

@typescript-eslint/explicit-module-boundary-types: Enforces explicit return types on functions and class methods. This helps improve code readability and provides better type safety in TypeScript.

@typescript-eslint/no-unused-vars: Flags unused variables, functions, and parameters in TypeScript code. This can help catch potential mistakes and clean up the code.

@typescript-eslint/no-explicit-any: Forbids the use of the any type in TypeScript code, promoting better type checking and avoiding potential bugs.

react-hooks/rules-of-hooks: Enforces rules of hooks when using React hooks. This helps prevent common mistakes and ensures proper hook usage.

react-hooks/exhaustive-deps: Warns when the dependency array in a React hook is not exhaustive, reducing the risk of stale closures.

eslint-plugin-react/jsx-props-no-spreading: Forbids the use of JSX spreading props ({...props}) to prevent unintended side effects and enforce more explicit prop passing.

eslint-plugin-react-native/no-inline-styles: Discourages the use of inline styles in React Native components, promoting better separation of concerns and maintainability.

eslint-plugin-import/no-extraneous-dependencies: Flags the use of unnecessary or uninstalled dependencies, helping to keep your dependencies tidy.

eslint-plugin-prefer-arrow: Encourages the use of arrow functions over regular functions in various scenarios, promoting concise syntax and avoiding issues with this binding.

eslint-plugin-prefer-template: Encourages the use of template literals instead of string concatenation, leading to more readable code.

**Comment by HaakonSvane** [comment]
We had a discussion about how what a rollout should look like. 
Since many of our apps really haven't cared about code rules for a good couple of years now, they are probably going to go all red when we apply our desired set of rules to them.
Tormod and Daniel suggested creating tiers of our config, much like what airbnb does for theirs (airbnb/base, airbnb/somethingElse), and let the implementing apps iteratively migrate.

#140 tracks the task of adding a linting system and there has been made a branch with some initial work added to it. It would be great if someone picked that up and continued where I left that off, keeping the things mentioned in this thread in mind.

**Comment by tormodAase** [comment]
https://github.com/Intellicode/eslint-plugin-react-native
![image](https://github.com/equinor/mad/assets/22524448/d2680e50-2dde-4b12-8473-bd321d28a0ba)

`no-unused-styles` 👀 
@dbjoEq @HaakonSvane @dansvend

**Comment by HaakonSvane** [comment]
I have a draft PR open here: #161 
We need to set some more rules as this currently only exposes the bare minimum `recommended` from some of the packages i ended up including (most from Daniel's list).

I have split up the configs into multiple parts:
- `@equinor/mad/base`
- `@equinor/mad/react`
- `@equinor/mad/typescript`

This means we can selectively extend any combination of these in our own projects as we see fit to avoid a too drastic change as we migrate. This system also works nice for projects that are not react projects for example, as you'd just extend `base` and `typescript` then.
You can, of course, reference all of them by just extending `@equinor/mad`
